### ISSUE TITLE
Make Material return 1 argument

### ISSUE THREAD

**Issue by MysteryPancake** [Facepunch/garrysmod-requests#999]
At the moment, Material returns 2 arguments, the IMaterial as well as the time taken. This makes Material impossible to use in a table:
```
local images = { Material( "icon16/coins.png" ), Material( "icon16/star.png" ), Material( "icon16/heart.png" ) }
PrintTable( images )
-- 1	=	Material [icon16/coins]
-- 2	=	Material [icon16/star]
-- 3	=	Material [icon16/heart]
-- 4	=	1.9379468882248e-06
```
It would make much more sense to remove this last argument, as I can't imagine it was ever used in the first place.

**Comment by FPtje** [comment]
Any function with multiple return values has that property. That's not an argument against `Material` specifically as much as it is one against Lua as a language. 

Basically, `{<some expression>, <some  other expression>}` might give you a table with **more** than two values. That's surprising, but not a GMod bug.

That said, the second return value of `Material` is weird on its own. Here's the description on the wiki:
```
Returns
  IMaterial
    Generated material
  number
    How long it took for the function to run
```

Surely if you want that timing information, you'd time it yourself, like you would with literally any other function. What is this second return value useful for?

**Comment by meepdarknessmeep** [comment]
It's useful to find if it is an error, if it was cached, or if it was a new material

**Comment by FPtje** [comment]
Then why doesn't it return that information instead? Why should that have to go through such an indirect method?

**Comment by MysteryPancake** [comment]
If the IMaterial is an error, then it should return an error for the first argument anyway.

**Comment by meepdarknessmeep** [comment]
It's just how Materials works I believe, if you change it you would needlessly be removing data that could be potentially useful anyways, just isolate the first argument yourself with parenthesis.

**Comment by MysteryPancake** [comment]
These 2 do literally print the exact same thing:
```
local time = SysTime()
local mat, time2 = Material( "icon16/star.png" )
print( SysTime() - time, time2 )
```
I'm sick of my tables being sacrificed for this

**Comment by meepdarknessmeep** [comment]
yes then look how inaccurate that is

**Comment by MysteryPancake** [comment]
Yeah, I just checked. They looked similar when I tried it for the first time. A separate function might be useful for this sort of thing though. I have a lot of tables with lots of materials, and it's a bit annoying to have to put brackets everywhere

**Comment by Kefta** [comment]
That's because Material also does stuff in the Lua state. It'd be more accurate if there wasn't an override.

**Comment by MysteryPancake** [comment]
I have an idea. What about 2 [new functions for IMaterials](https://wiki.garrysmod.com/page/Category:IMaterial) that lets you get the time taken to create them, and whether or not they are cached?
You can already tell if they are an error with [IMaterial:IsError](https://wiki.garrysmod.com/page/IMaterial/IsError).

**Comment by FPtje** [comment]
> yes then look how inaccurate that is

Of *course* it's inaccurate. That's the entire point. Using the time something took to figure out whether it did something is *utterly ridiculous*. That applies to measuring it in Lua **AND** using the second return value of `Material`.

It could have returned a boolean saying whether it was cached or not. With that you'd be 100% sure and 100% accurate in knowing whether a material was cached or not. Hell, there could be a function that returns whether a specific material is cached.

*Surely* that return value has a more sensible purpose?

**Comment by Kefta** [comment]
I'm nearly certain it was just leftover from Garry or Rubat debugging. Obviously there's no purpose beyond that.

**Comment by FPtje** [comment]
That would make a *lot* more sense.

**Comment by Tripperful** [comment]
@Kefta, though this second returned argument is documented on the wiki, so it's probably not a "debug leftover". I support the idea of just getting rid of this second returned value as it brings more pain in the ass than any useful data. It looks like some hack, why Material is the only function behaving this way? Why don't Model(), Sound() or any other caching functions return such a value? If you want to profile the time elapsed for something, use time stamps before and after the function call

**Comment by Kefta** [comment]
Benchmarking it in C is much more accurate than in Lua. It's probably there because Material creates an actual object while Model and Sound are just Lua shortcuts to util.Precache*. It's probably documented on the wiki because it was LEFT IN for debugging, and so that no one was confused about what it was.

**Comment by Tripperful** [comment]
@Kefta, I see your point. I mean that I don't see any use for this super-precise benchmark of some single arbitrary function in GLua environment. It could be somewhat useful in some more low-level places, but really, who have ever used this argument of Material?

**Comment by Kefta** [comment]
I'm not arguing that it should remain in. My point is that people are making vastly wild assertions about the extra argument when in reality, it's obviously there for debugging for Garry, Rubat, or whoever. It's also extremely easy to work around and I don't understand the severity of tone in working with the function.

**Comment by MysteryPancake** [comment]
@meepdarknessmeep I just wanted to say, I don't think Material gets cached anyway. Every tutorial I've ever read says to never use Material in a HUDPaint loop since the material is recreated every frame. Also, there's no possible way to tell if a material was cached or not anyway just using a time. It obviously differs on different systems, so it can't possibly be accurate enough.

**Comment by meepdarknessmeep** [comment]
it gets cached but that doesnt mean more calls creates more overhead/memory

**Comment by ZenkakuHiragana** [comment]
Perhaps it's too late to post, but it can be easily workarounded by adding `nil` at the end.

```Lua
local function returnTwoValues(arg)
    return arg, -1
end

local a = { returnTwoValues("alfa"), returnTwoValues("bravo"), returnTwoValues("charlie") }
PrintTable(a)
print("")
-- 1	=	alpha
-- 2	=	bravo
-- 3	=	charley
-- 4	=	-1
--

local b = { returnTwoValues("alfa"), returnTwoValues("bravo"), returnTwoValues("charlie"), nil }
PrintTable(b)
-- 1	=	alpha
-- 2	=	bravo
-- 3	=	charley
```

**Comment by Kefta** [comment]
Or just wrap the multi-return call in extra parentheses.

**Comment by MysteryPancake** [comment]
Just got a notification for this, wow this is a throwback, hi Kefta! I was very young and silly when I opened this issue, definitely better to keep it as is :)
